[FLINK-3234] [dataSet] Add KeySelector support to sortPartition operation.

chiwanpark opened the pull request:
This PR contains following changes:

* Add `sortPartition` methods which receive a `KeySelector` instance or a Scala lambda function
* Add `SortPartitionOperator` constructor which takes `Keys.SelectorFunctionKeys`
* Add `getFlatFields` method to `SortPartitionOperator` class for `Keys.SelectorFunctionKeys`
* Add some test cases

fhueske commented on flink-java/src/main/java/org/apache/flink/api/java/DataSet.java:
```diff
@@ -1377,6 +1377,20 @@ public long count() throws Exception {
 	}
 
+	/**
+	 * Locally sorts the partitions of the DataSet on the specified field in the specified order.
```
"on the specified field" -> "on an extracted key"

fhueske commented on flink-java/src/main/java/org/apache/flink/api/java/DataSet.java:
```diff
@@ -1377,6 +1377,20 @@ public long count() throws Exception {
 
+	/**
+	 * Locally sorts the partitions of the DataSet on the specified field in the specified order.
+	 * DataSet can be sorted on multiple fields by chaining sortPartition() calls.
```
I would not allow chaining of KeySelectors. We only allow a single key extractor anywhere else in the DataSet API (for a good reason). If users want to sort on more than one field, they can return a `TupleX` from the `KeySelector`.

fhueske commented on flink-java/src/main/java/org/apache/flink/api/java/operators/SortPartitionOperator.java:
```diff
@@ -59,6 +62,14 @@ public SortPartitionOperator(DataSet<T> dataSet, String sortField, Order sortOrd
+		super(dataSet, dataSet.getType());
+		this.sortLocationName = sortLocationName;
+
+		int[] flatOrderKeys = getFlatFields(sortKey);
```
This does not work. 
`SelectorFunctionKeys` need to be handled completely different. In short, you need to inject a MapFunction<IN, Tuple2<KEY, IN>> that extracts the key and emits a Tuple2 with the extracted key and the original input record.
The data set is then sorted on the first tuple field (the key field) and the records are unwrapped after the sort by another `MapFunction<Tuple2<KEY, IN>, IN>`. Have a look at how the PartitionOperator handles KeySelectors.

fhueske commented on flink-java/src/main/java/org/apache/flink/api/java/operators/SortPartitionOperator.java:
```diff
@@ -89,6 +100,22 @@ public SortPartitionOperator(DataSet<T> dataSet, String sortField, Order sortOrd
+	 * @param order The order of the additional sort order of the local partition sorting.
+	 * @return The DataSet with sorted local partitions.
+	 */
+	public <K> SortPartitionOperator<T> sortPartition(KeySelector<T, K> keyExtractor, Order order) {
```
I would not allow chaining.
In addition, we must make sure that `ExpressionKeys` and `SelectorFunctionKeys` are not mixed. So you can either use a single key selector or one or more expression keys.

fhueske commented on flink-java/src/test/java/org/apache/flink/api/java/operator/SortPartitionTest.java:
```diff
@@ -169,6 +169,38 @@ public void testSortPartitionWithExpressionKeys4() {
+			tupleDs.sortPartition(new KeySelector<Tuple4<Integer, Long, CustomType, Long[]>, Integer>() {
+				@Override
+				public Integer getKey(Tuple4<Integer, Long, CustomType, Long[]> value) throws Exception {
+					return value.f0;
```
The test only works because the KeySelector function returns the first field of the input data. If you change this to `value.f1` the test will fail.

fhueske replied:
Can you return some other field than `f0` here?

fhueske commented on flink-java/src/test/java/org/apache/flink/api/java/operator/SortPartitionTest.java:
```diff
@@ -169,6 +169,38 @@ public void testSortPartitionWithExpressionKeys4() {
 		tupleDs.sortPartition("f3", Order.ASCENDING);
 	}
 
+	@Test
```
I suggest add more tests:
- mixing `SelectorFunctionKeys` and `ExpressionKeys` -> Should not work
- use a `KeySelector` that returns a composite type, e.g., a `Tuple3`

fhueske commented on flink-scala/src/main/scala/org/apache/flink/api/scala/PartitionSortedDataSet.scala:
```diff
@@ -39,12 +41,25 @@ class PartitionSortedDataSet[T: ClassTag](set: SortPartitionOperator[T])
+  /**
+    * Appends the given field and order to the sort-partition operator.
+    */
+  override def sortPartition[K: TypeInformation](fun: T => K, order: Order): DataSet[T] = {
```
Do not allow chaining of `KeySelector`s. Make sure different key types are not mixed.

fhueske commented on flink-tests/src/test/scala/org/apache/flink/api/scala/operators/SortPartitionITCase.scala:
```diff
@@ -166,6 +166,23 @@ class SortPartitionITCase(mode: TestExecutionMode) extends MultipleProgramsTestB
   }
 
+  @Test
+  def testSortPartitionWithKeySelector(): Unit = {
```
Add same tests as for Java.

fhueske commented:
Hi @chiwanpark, thanks for picking up this issue and filling this gap in the DataSet API. 
Would be nice, if it could make it into the 1.0 release, IMO.
However, the implementation needs to be improved. I suggest to have a look at the `PartitionOperator` and check how `KeySelector`s are handled there.

chiwanpark commented:
Thanks for review @fhueske! I'll address your comments.

chiwanpark commented:
@fhueske I addressed your comments.

To reduce modifying original code, I added some variables for key selector only. But this make the source code quite messy. I would like to refactor `SortPartitionOperator` like following:

* Preserve key informations and sort orders as `List<Keys>` and `List<Order>` object.
  * We have to create a list because users can chain expression keys.
* Check whether given field is sortable or not in constructors and `sortPartition` method in `SortPartitionOperator`
* Remove `sortKeyPositions`, `sortOrders`, `getFlatFields` and `appendSorting`
* When `translateToDataFlow` method is called, the method fetches key positions from preserved key objects and creates operators.

Does this make sense?

fhueske commented on flink-java/src/main/java/org/apache/flink/api/java/DataSet.java:
```diff
@@ -1377,6 +1377,20 @@ public long count() throws Exception {
 
+	/**
+	 * Locally sorts the partitions of the DataSet on the an extracted key in the specified order.
+	 * DataSet can be sorted on multiple values by returning a tuple.
```
"returning a tuple" + "from the KeySelector."

fhueske commented on flink-java/src/main/java/org/apache/flink/api/java/operators/SortPartitionOperator.java:
```diff
@@ -79,16 +112,33 @@ public SortPartitionOperator(DataSet<T> dataSet, String sortField, Order sortOrd
+	 * @param order        The order of the additional sort order of the local partition sorting.
+	 * @return The DataSet with sorted local partitions.
+	 */
+	public <K> SortPartitionOperator<T> sortPartition(KeySelector<T, K> keyExtractor, Order order) {
```
If you remove this method, it cannot be called ;-)

chiwanpark replied:
If we remove this method, the following code can be executed:

```java
DataSet<MyObject> data = ...
DataSet<MyObject> result = data
  .sortPartition(new KeySelector<MyObject, Integer> {
    public Integer getKey(MyObject value) throws Exception {
      return value.myInt;
    }
  }, Order.ASCENDING)
  .sortPartition(new KeySelector<MyObject, String> {
    public String getKey(MyObject value) throws Exception {
      return value.myString;
    }
  }, Order.ASCENDING);
```

In above case, `data` is sorted twice (first by Integer, second by String) and the result of first sorting will be ignored. I think this is quite confusing for users.

fhueske replied:
Oh yes, you are right! 
Can you update the JavaDocs of this method and explain that chaining is not possible and how to work around it?
Thanks

fhueske commented on flink-scala/src/main/scala/org/apache/flink/api/scala/PartitionSortedDataSet.scala:
```diff
@@ -35,16 +37,30 @@ class PartitionSortedDataSet[T: ClassTag](set: SortPartitionOperator[T])
     this
   }
 
+  override def sortPartition[K: TypeInformation](fun: T => K, order: Order): DataSet[T] = {
```
remove method.

fhueske commented on flink-tests/src/test/java/org/apache/flink/test/javaApiOperators/SortPartitionITCase.java:
```diff
@@ -197,6 +198,58 @@ public void testSortPartitionParallelismChange() throws Exception {
+			.sortPartition(new KeySelector<Tuple3<Integer, Long, String>, Integer>() {
+				@Override
+				public Integer getKey(Tuple3<Integer, Long, String> value) throws Exception {
+					return value.f0;
```
return some other field than `f0`.

fhueske commented:
Hi @chiwanpark, thanks for the update!
The refactorings you proposed make a lot of sense. +1

Thanks, Fabian

fhueske commented on flink-java/src/test/java/org/apache/flink/api/java/operator/SortPartitionTest.java:
```diff
@@ -169,6 +169,72 @@ public void testSortPartitionWithExpressionKeys4() {
+		}, Order.ASCENDING);
+	}
+
+	@Test(expected = InvalidProgramException.class)
```
Can you add a test that first uses KeySelectors and then ExpressionKeys?

fhueske commented on flink-java/src/main/java/org/apache/flink/api/java/DataSet.java:
```diff
@@ -1377,6 +1377,24 @@ public long count() throws Exception {
 	}
 
+	/**
+	 * Locally sorts the partitions of the DataSet on the an extracted key in the specified order.
```
"...the DataSet on the **an** extracted key...", remove "an"

fhueske commented on flink-java/src/main/java/org/apache/flink/api/java/DataSet.java:
```diff
@@ -1377,6 +1377,24 @@ public long count() throws Exception {
 
+	/**
+	 * Locally sorts the partitions of the DataSet on the an extracted key in the specified order.
+	 * DataSet can be sorted on multiple values by returning a tuple from the KeySelector.
```
"The DataSet can be ...", add "The"

fhueske commented on flink-java/src/main/java/org/apache/flink/api/java/DataSet.java:
```diff
@@ -1377,6 +1377,24 @@ public long count() throws Exception {
+	 * Locally sorts the partitions of the DataSet on the an extracted key in the specified order.
+	 * DataSet can be sorted on multiple values by returning a tuple from the KeySelector.
+	 *
+	 * Note that any key extraction methods cannot be chained with the KeySelector. To sort the
```
"Note that any key extraction methods cannot be ..." -> "Note that no additional sort keys can be appended to a KeySelector."

fhueske commented on flink-java/src/main/java/org/apache/flink/api/java/operators/SortPartitionOperator.java:
```diff
@@ -36,27 +40,58 @@
+		orders.add(sortOrder);
+	}
+
+	public SortPartitionOperator(DataSet<T> dataSet, Keys<T> sortKey, Order sortOrder, String sortLocationName) {
```
Change the `sortKey` parameter type to `SelectorFunctionKeys` (or accept the `KeySelector` and create the `SelectorFunctionKeys` in the constructor.

fhueske commented on flink-java/src/main/java/org/apache/flink/api/java/operators/SortPartitionOperator.java:
```diff
@@ -79,58 +119,41 @@ public SortPartitionOperator(DataSet<T> dataSet, String sortField, Order sortOrd
-				this.sortKeyPositions[oldLength+i] = flatOrderFields[i];
-				this.sortOrders[oldLength+i] = order;
-			}
+	private void ensureSortableKey(Keys<T> sortKey) {
```
Change the `sortKey` parameter type to `SelectorFunctionKeys` and remove the type check + cast.

fhueske commented on flink-scala/src/main/scala/org/apache/flink/api/scala/DataSet.scala:
```diff
@@ -1508,6 +1508,31 @@ class DataSet[T: ClassTag](set: JavaDataSet[T]) {
+    * partition by multiple values using KeySelector, the KeySelector must return a tuple
+    * consisting of the values.
+    */
+  def sortPartition[K: TypeInformation](fun: T => K, order: Order): DataSet[T] ={
```
Copy the method docs from the `DataSet.java`.

fhueske commented on flink-tests/src/test/java/org/apache/flink/test/javaApiOperators/SortPartitionITCase.java:
```diff
@@ -197,6 +198,58 @@ public void testSortPartitionParallelismChange() throws Exception {
+				public Long getKey(Tuple3<Integer, Long, String> value) throws Exception {
+					return value.f1;
+				}
+			}, Order.DESCENDING)
```
Change sort order to `ASCENDING` (or in the other test).

fhueske commented on flink-tests/src/test/scala/org/apache/flink/api/scala/operators/SortPartitionITCase.scala:
```diff
@@ -166,6 +167,58 @@ class SortPartitionITCase(mode: TestExecutionMode) extends MultipleProgramsTestB
+
+    val result = ds
+      .map { x => x }.setParallelism(4)
+      .sortPartition(_._2, Order.DESCENDING)
```
Change sort order to `ASCENDING` (or in the other test).

fhueske commented:
The refactoring looks good, @chiwanpark.
I have just a few minor remarks. The PR can be resolved after these have been addressed.

chiwanpark commented:
Thanks for review @fhueske. I have addressed your comments. :-)

fhueske commented:
Thanks for the fast update!
Good to merge :-)

fhueske commented:
Will merge this PR.
